### PR TITLE
Pylint tree tests

### PR DESCRIPTION
**Pull request by RAMitchell**
- Remove tests path from pytest.ini so we can run individual tests instead of running all tests every time 

- Reduce the duration of a few longer running tree tests

- Add .pylintrc and lint tests/explainers/test_tree.py

- Tests that fail optional imports now get marked as skipped instead of succeeded

- If pyspark cluster cannot be instantiated, mark as skipped instead of succeeded

### CODE REVIEW

File: tests/explainers/test_tree.py [hunk 1]
```diff
@@ -452,39 +402,29 @@ def test_catboost():
...
-    y_val = train_df.ACTION[ix:ix+20]
-    model = catboost.CatBoostClassifier(iterations=100, learning_rate=0.5, random_seed=12)
+        "SHAP values don't sum to modThisel output!"
+
+    X, y = sklearn.datasets.load_breast_cancer(return_X_y=True)
+    model = catboost.CatBoostClassifier(iterations=10, learning_rate=0.5, random_seed=12)
     model.fit(
-        X_train,
-        y_train,
-        eval_set=(X_val, y_val),
```
Comment (lrjball) [inline tests/explainers/test_tree.py:467]: Any reason for eval_set being removed here?

File: tests/explainers/test_tree.py [hunk 2]
```diff
@@ -610,37 +530,29 @@ def test_lightgbm_binary():
...
+    X_train, _, Y_train, _ = sklearn.model_selection.train_test_split(*shap.datasets.iris(),
+                                                                      test_size=0.2, random_state=0)
+    rforest = sklearn.ensemble.RandomForestClassifier(n_estimators=100, max_depth=None,
+                                                      min_samples_split=2,
+                                                      random_state=0)
     model = rforest.fit(X_train, Y_train)
 
     # verify symmetry of the interaction values (this typically breaks if anything is wrong)
     interaction_vals = shap.TreeExplainer(model).shap_interaction_values(X)
-    for i in range(len(interaction_vals)):
```
Comment (lrjball) [inline tests/explainers/test_tree.py:627]: Is this change necessary? Feels like range(len) is already fine here. Same for below

File: pytest.ini [hunk 3]
```diff
@@ -1,2 +1,2 @@
 [pytest]
-addopts = --mpl --cov=shap tests --cov-report=term-missing
\ No newline at end of file
+addopts = --mpl --cov=shap --cov-report=term-missing --durations=20
```
Comment (lrjball) [inline pytest.ini:2]: Is this causing timeout issues in the CI? Don't think we want to skip tests by default in the pytest.ini Maybe that setting could be applied by a user if the tests are taking too long, or in the offending CI config file if the CI is failing

File: .travis.yml [hunk 4]
```diff
@@ -18,7 +18,9 @@ install:
 - python setup.py develop
 - pip install .[test]
 - pip install tensorflow xgboost matplotlib ipython codecov torch torchvision
-script: python -m pytest .
+script:
+  - python -m pytest tests
+  - pylint tests/explainers/test_tree.py
```
Comment (lrjball) [inline .travis.yml:23]: Has this been left in accidentally? I guess we would either lint the whole project as part of the CI or none of it, rather than just this one file

**Review by lrjball — COMMENTED** [review]
Couple of small points raised, but other than that LGTM

File: tests/explainers/test_tree.py (hunk 1 above)
Comment (RAMitchell) [reply tests/explainers/test_tree.py:467]: It has no purpose and takes time to evaluate.

File: tests/explainers/test_tree.py (hunk 2 above)
Comment (RAMitchell) [reply tests/explainers/test_tree.py:627]: This one is a pylint failure. I agree that there is not much difference, but IMO it's better to conform.

File: pytest.ini (hunk 3 above)
Comment (RAMitchell) [reply pytest.ini:2]: This only prints the durations of the 20 longest tests.

File: .travis.yml (hunk 4 above)
Comment (RAMitchell) [reply .travis.yml:23]: I spent about 3 hours conforming this one file, because I will be working on it. It's just not possible to do the entire code base in one go, It's too much work and the diff is too large.

**Comment by RAMitchell** [comment]
@slundberg

File: .travis.yml (hunk 4 above)
Comment (lrjball) [reply .travis.yml:23]: I wasn't suggesting you reformat the entire code base, don't worry! One thing I would say though is that there may be better ways to contribute than to reformat all of the test files. Ultimately formatting is a choice and can often lead to a lot of debate, but doesn't add too much value to the code base one way or another. 

I think adding test skips is definitely worthwhile, and moving imports too, but other than that it is just another potential source of a bug creeping into code which already works.

File: pytest.ini (hunk 3 above)
Comment (lrjball) [reply pytest.ini:2]: Ah my bad, good shout!

File: tests/explainers/test_tree.py (hunk 2 above)
Comment (lrjball) [reply tests/explainers/test_tree.py:627]: pylint is only really a list of guidelines, and doesn't necessarily have to be followed religiously. Remember that every change takes time both to make, and to have someone else review, so if there is no difference either way my opinion would always be to leave it how it is.

File: tests/explainers/test_tree.py (hunk 1 above)
Comment (lrjball) [reply tests/explainers/test_tree.py:467]: I guess including it does some basic testing of eval_set to make sure that nothing errors when it is set, but this could definitely be improved by adding some further testing to make sure eval_set behaves as intended.

**Review by slundberg — APPROVED** [review]
Thanks @RAMitchell ! This looks like a good start to separating out the code to support GPU testing. I had orgininally left the imports inside each function to make it easy to copy and paste them into a notebook for debugging, but I think moving them to the top is a good choice.

I expect there may be some linting changes for the future, but happy have this file conform as a start.

Thanks for taking a look @lrjball !
